Rubberduck.Parsing/Rewriter/SelectionRecoverer.cs

```diff
@@ -0,0 +1,181 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Rubberduck.Parsing.VBA;
+using Rubberduck.VBEditor;
+using Rubberduck.VBEditor.Extensions;
+using Rubberduck.VBEditor.Utility;
+
+namespace Rubberduck.Parsing.Rewriter
+{
+    public sealed class SelectionRecoverer : ISelectionRecoverer, IDisposable
```

Comment:
Is this guaranteed to only be called from a single thread?

Reply:
No. Why do you ask?

Currently, this is only used by the rewrite sessions.

Reply:
My main concern was the thread safety of the caches vis-a-vis `SaveSelections` and other callers. If it's only being called by rewrite sessions, that _should_ be fine.

Reply:
Note thatit is registered in transient scope. So, all users will get their own cache. Well, the rewrite sessions all get the same one via their fatcory, but they are guaranteed to never execute in parallel.